File: shell/packages/app/src/auth/slug.test.ts
import { describe, expect, it } from "vitest";
import { deriveSlug, isValidSlug } from "./slug.js";

// The same cases as the engine's TestDeriveSlug, so the two stay in step.
describe("deriveSlug", () => {
  it.each([
    ["Acme Corp", "acme-corp"],
    ["  Acme   Corp  ", "acme-corp"],
    ["Café Ghana Ltd.", "cafe-ghana-ltd"],
    ["O'Brien & Sons", "o-brien-sons"],
    ["3M Ghana", "co-3m-ghana"],
    ["--Acme--", "acme"],
    ["安全", ""],
    ["AB", "ab"],
    ["abc-".repeat(30), "abc-".repeat(16).replace(/-+$/, "")],
  ])("derives %j as %j", (name, want) => {
    expect(deriveSlug(name)).toBe(want);
  });
});

describe("isValidSlug", () => {
  it.each([
    ["Acme Corp", true],
    ["3M Ghana", true],
    ["AB", false],
    ["安全", false],
    ["abc-".repeat(30), true],
  ])("reports %j's slug valid: %j", (name, want) => {
    expect(isValidSlug(deriveSlug(name))).toBe(want);
  });
});

File: internal/engine/auth/authregister/slug_test.go
package authregister

import (
	"strings"
	"testing"
)

func TestDeriveSlug(t *testing.T) {
	cases := []struct{ name, want string }{
		{"Acme Corp", "acme-corp"},
		{"  Acme   Corp  ", "acme-corp"},
		{"Café Ghana Ltd.", "cafe-ghana-ltd"},
		{"O'Brien & Sons", "o-brien-sons"},
		{"3M Ghana", "co-3m-ghana"},
		{"--Acme--", "acme"},
		{"安全", ""},
		{"AB", "ab"},
		{strings.Repeat("abc-", 30), strings.TrimRight(strings.Repeat("abc-", 16), "-")},
	}
	for _, c := range cases {
		if got := DeriveSlug(c.name); got != c.want {
			t.Errorf("DeriveSlug(%q) = %q, want %q", c.name, got, c.want)
		}
	}
}

func TestDeriveSlug_ValidityFollowsTheSlugRule(t *testing.T) {
	for name, want := range map[string]bool{
		"Acme Corp":                true,
		"3M Ghana":                 true,
		"AB":                       false,
		"安全":                       false,
		strings.Repeat("abc-", 30): true,
	} {
		if got := validSlug(DeriveSlug(name)); got != want {
			t.Errorf("validSlug(DeriveSlug(%q)) = %v, want %v", name, got, want)
		}
	}
}
